src/metabase/driver/sql_jdbc/execute.clj

```diff
@@ -232,19 +232,29 @@
             (finally
               (throw e))))))))
 
+(defn- prepare-result
+  "Convert a result set to a hash-map with :rows and :columns."
+  [[columns & rows]] {:rows (or rows []) :columns (map u/qualified-name columns)})
+
 (defn- run-query
   "Run the query itself."
-  [driver {sql :query, :keys [params remark max-rows]} connection]
+  [driver {sql :query, :keys [params remark max-rows data-fn]}, connection]
   (let [sql              (str "-- " remark "\n" sql)
-        [columns & rows] (cancelable-run-query
-                          connection sql params
-                          {:identifiers    identity
-                           :as-arrays?     true
-                           :read-columns   (partial read-columns driver)
-                           :set-parameters (partial set-parameters driver)
-                           :max-rows       max-rows})]
-    {:rows    (or rows [])
-     :columns (map u/qualified-name columns)}))
+        res              (cancelable-run-query
+                           connection sql params
+                           {:identifiers    identity
+                            :as-arrays?     true
+                            :result-set-fn  (if data-fn
```

Comment:
Does this `if` deliberately have no else clause? If so, how about making it a `when`?